Restructure polymer cli docs

katejeffreys opened the pull request:
Addresses #2071 in response to @FredKSchott's comments in #2063

katejeffreys commented:
Open to suggestions for other ways of doing this. Also needs a nav fix which I will do last.

FredKSchott commented on app/2.0/docs/tools/create-app-polymer-cli.md:
```diff
@@ -0,0 +1,102 @@
+<!-- toc -->
+
+Polymer CLI supports initializing a project folder with one of several application templates.  The 
+CLI comes with a `basic` template, which is the most basic starting point for a Polymer-based 
```
Confusing: Putting basic in ```s makes it feel like a proper name, but the actual term "basic" is never used by the CLI & there's no one "basic" template. 

Suggestion: "The CLI comes with a basic `polymer-2-app` template, which is the..."

FredKSchott commented on app/2.0/docs/tools/create-app-polymer-cli.md:
```diff
@@ -0,0 +1,102 @@
+
+## App project architecture {#app-architecture}
+
+Polymer CLI is designed for apps that follow the [app shell 
```
New user 🎩: Wait, Is it designed *only* for apps that follow the app shell architecture? If I'm not following that, i can't use the CLI?

FredKSchott commented on app/2.0/docs/tools/create-app-polymer-cli.md:
```diff
@@ -0,0 +1,102 @@
+
+There are fundamental concepts of the app shell architecture that you should understand before 
+creating your app project with Polymer CLI: the entrypoint, the shell, and fragments. See [App 
+structure](../../toolbox/prpl#app-structure) from the App Toolbox docs for an in-depth overview of 
```
🎩: I just read the developers.google.com link above, is this different? How? Do I need to read both?

FredKSchott commented on app/2.0/docs/tools/create-app-polymer-cli.md:
```diff
@@ -0,0 +1,102 @@
+
+## Set up basic app project {#basic-app}
+
+Follow the steps below to get your `basic` app project set up.
```
Okay I think I see now that you're using "`basic`" as a name for the static, not-app-shell applications. This isn't clearly defined anywhere I've seen, if we want to use this we should explicitly describe what a `basic` app is somewhere here, or elsewhere that we can link to.

arthurevans replied:
Yeah, this shouldn't be in backticks `basic` unless it's a literal string (like `polymer-2-application`).

arthurevans replied:
Still have the backticks.

FredKSchott commented on app/2.0/docs/tools/create-app-polymer-cli.md:
```diff
@@ -0,0 +1,102 @@
+
+1.  Select `polymer-2-application`.
+
+1.  Enter a name for your app. Defaults to the name of the current directory.
```
do we need to describe each step here? These steps are subject to change through PRs, it would be unreasonable to expect our entire team to remember to come back here and update each time.

Suggestion: Replace these 3 steps with a summary "follow the prompts to create your app" step. The prompts are meant to be easy to follow for beginners and should contain enough info for the beginning user.

katejeffreys replied:
I agree that it would be unreasonable to expect the team to remember to update the CLI docs with these changes. I will maintain this documentation :)

For now I'm happy to leave these steps here with explanations for each. For instance, the user is asked to give a name to their "main element" and as you've noted below, the concept of a "main element" is not clear without an explanation.

arthurevans replied:
If we think this is going to be changing rapidly, we could move to just describing the "main element" concept up front, like, "Follow the prompts to create your app. Among other things, the CLI prompts your for the main element in your project..." 

But why don't we go with the steps for now and revisit if they end up changing frequently?

FredKSchott commented on app/2.0/docs/tools/create-app-polymer-cli.md:
```diff
@@ -0,0 +1,102 @@
+
+1.  Enter a description for your app.
+
+At this point, Polymer CLI generates files and directories for your app, and installs your 
```
Suggestion: "the Polymer CLI generates your app and installs its dependencies."

FredKSchott commented on app/2.0/docs/tools/create-app-polymer-cli.md:
```diff
@@ -0,0 +1,102 @@
+
+## App project layout
+
+After the initialization process Polymer CLI generates the following files and directories.
```
Suggestion: "initialization process" feels a little 🤖

FredKSchott commented on app/2.0/docs/tools/create-app-polymer-cli.md:
```diff
@@ -0,0 +1,102 @@
+
+### Add elements
+
+You may want to compose your main element out of smaller, application-specific elements. These 
```
🎩: What's a main element? Where is that? What's an application-specific element?

FredKSchott commented on app/2.0/docs/tools/create-app-polymer-cli.md:
```diff
@@ -0,0 +1,102 @@
+        my-app/
+        my-el/
+
+Currently there is no Polymer CLI command to generate application-specific elements. You should do 
```
No need to highlight a shortcoming of the CLI/experience if we don't need to :) 

Suggestion: Remove, or replace with something like "Polymer applications can be composed of many elements in your `src/` folder"

arthurevans replied:
I think we kind of need to highlight this to prevent people from trying to do init element inside an app project. But instead of saying, "There is currently no..." We could say,

To add another element to the project:

1. Create a new folder under src.

        mkdir src/my-el

2. Create an HTML import for the new element. You can use the existing app element as a starting point.

3. To use the new element, you'll need to import it into your application element.

 Don't use `polymer init` to create an element project inside your app project.

This adds more detail without being entirely satisfying (i.e., doesn't show adding the `<link rel="import">` or using the element in the app element). So alternately, just add the .html files to the diagram above for clarity:


```
my-example/
     src/
       my-example-app/
         my-example-app.html
       my-el/
         my-el.html
```

You can copy the existing app element as a starting point for your next element.

 Don't use `polymer init` to create an element project inside your app project.

=== end suggestion

I changed the example here to "my-example" so the directory names would make sense. If you're using `my-app` as a placeholder, should be `<var>my-app</var>` by style. Downside: examples might need to be formatted in HTML instead of fenced code blocks, because fenced code blocks escape the HTML markup.

FredKSchott commented on app/2.0/docs/tools/create-app-polymer-cli.md:
```diff
@@ -0,0 +1,102 @@
+    <!-- from a 'src/some-application.html' file in your application -->
+    <link rel="import" href="../bower_components/polymer/polymer.html">
+
+Only one file should use absolute URLs for imports: your "entrypoint" (defaults to "index.html"). 
```
🎩: wait, what defaults to "index.html"?

Suggestion: `(default: "index.html")`

FredKSchott commented on app/2.0/docs/tools/create-element-polymer-cli.md:
```diff
@@ -0,0 +1,121 @@
+
+This section shows you how to start an element project.
+
+1.  Create a directory for your element project. It's best practice for the name
```
Not sure if the "It's best practice" line is necessary / overall directory naming strategies are probably out of scope for this doc. 

If you do keep it, then explain why its a best practice.

FredKSchott commented on app/2.0/docs/tools/create-element-polymer-cli.md:
```diff
@@ -0,0 +1,121 @@
+Dependencies are stored in the `bower_components` directory. You should never manually alter the
+contents of this directory.
+
+Use the Bower CLI to manage dependencies.
```
🎩 what about that `polymer install` command?

arthurevans replied:
This is a good question. Don't want to hang up on it too much here (we use bower directly other places), but we should discuss whether to move to that in general (or offer both options).

FredKSchott commented on app/2.0/docs/tools/create-element-polymer-cli.md:
```diff
@@ -0,0 +1,121 @@
+
+Polymer CLI uses [Bower](http://bower.io) for dependency management.
+
+Dependencies are stored in the `bower_components` directory. You should never manually alter the
```
How bower works is probably out of scope for this document as well.

arthurevans replied:
Since we're just about the only folks who use it, seems like it's probably OK. If you're comfortable with Bower and CLI tools, chances are you aren't even reading the "create an element project/app project" docs. I wouldn't want to give a full tutorial, but a few "don't stick your fork in this here socket" pointers seem apropos?

FredKSchott commented on app/2.0/docs/tools/create-element-polymer-cli.md:
```diff
@@ -0,0 +1,121 @@
+
+Use the Bower CLI to manage dependencies.
+
+This command downloads a dependency to `bower_components/` (the `--save` flag saves the new 
```
Which command?

(also, see above about Bower behavior probably being out of scope for this doc)

arthurevans replied:
Rather than "This command" form, maybe replace these with:

To add a new Bower dependency to your project:

    <bower command>

As Fred suggested, "This command..." is a little ambiguous about whether you're talking about the command above, or the command below.

FredKSchott commented on app/2.0/docs/tools/create-element-polymer-cli.md:
```diff
@@ -0,0 +1,121 @@
+
+    bower install --save PolymerElements/iron-ajax
+
+This command removes the dependency from `bower_components` and `bower.json`:
```
Which command?

FredKSchott commented on app/2.0/docs/tools/create-element-polymer-cli.md:
````diff
@@ -0,0 +1,121 @@
+<link rel="import" href="../paper-elements/paper-button.html">
+```
+
+Using relative URLs may feel weird, but remember that dependencies are siblings of each other. So 
````
highlight exactly why it feels weird, in case a user doesn't know why its weird or if a user is feeling weird but unsure if its for the right reason.

Also, "remember that dependencies are siblings of each other *once installed*".

FredKSchott commented on app/2.0/docs/tools/create-element-polymer-cli.md:
```diff
@@ -0,0 +1,121 @@
+
+#### Why use relative URLs?
+
+Suppose that you ran Polymer CLI to generate an element project. Your element is named `my-el`. You 
```
You already explain this above. No need to explain in two places.

katejeffreys commented:
Thanks for the suggestions @FredKSchott! I have revised the pages and am happy to answer questions about any of the choices.

After review from @arthurevans this can be merged!

arthurevans commented:
Do you have this staged anywhere? Having trouble visualizing.

arthurevans commented on app/2.0/docs/tools/create-app-polymer-cli.md:
```diff
@@ -0,0 +1,101 @@
+## App project architecture {#app-architecture}
+
+The `polymer-2-application` template follows the [app shell 
+architecture](https://developers.google.com/web/updates/2015/11/app-shell) pattern.
```
I think this whole section is incorrect, or at least misleading. The `polymer-2-application` template doesn't really have any structure at all. The PSK template follows the app shell architecture.

arthurevans replied:
Still true.

arthurevans commented on app/2.0/docs/tools/polymer-cli-commands.md:
```diff
@@ -0,0 +1,278 @@
+Generates a production-ready build of your app. This process includes minifying the HTML, CSS, and
+JS of the application dependencies, and generating a service worker to pre-cache dependencies.
+
+Polymer CLI's build process is designed for apps that follow the [app shell
```
I think this link is a little confusing because some of our app structure is Polymer- or PRPL-specific (Like, the terms "entrypoint" and "fragment"). And that doc is full of a lot of implementation details that won't be helpful to the average Polymer user.

It would make more sense to link here to the PRPL doc, and have only that doc link to the app shell article for background on the app shell pattern.

arthurevans commented on app/2.0/docs/tools/polymer-cli.md:
```diff
@@ -39,500 +39,28 @@ You're all set. Run `polymer help` to view a list of commands.
-Polymer CLI is a command-line interface for Polymer projects. It includes a build pipeline, a
-boilerplate generator for creating elements and apps, a linter, a development server, and a test
-runner.
+Polymer CLI is our official command line tool for Polymer projects and Web Components. It includes 
```
our => the (see style/person)

arthurevans commented on app/2.0/docs/tools/polymer-cli.md:
```diff
@@ -39,500 +39,28 @@ You're all set. Run `polymer help` to view a list of commands.
 
 ## Overview {#overview}
```
I know this organization is not new, but it strikes me as strange that the overview follows the install. 

Maybe move the first sentence to the head of the article (before the Install heading) and retitle this section "Types of projects" or "CLI project types".

arthurevans replied:
And maybe `## Install` => `## Install Polymer CLI`?

Not sure. Do what feels right.

arthurevans commented on app/2.0/docs/tools/polymer-cli-commands.md:
```diff
@@ -0,0 +1,278 @@
+* [`polymer test`](#tests)
+* [Global options for Polymer CLI commands](#global)
+
+## `polymer build` {#build}
```
As discussed, remove code formatting in headers.

katejeffreys commented:
Thanks, updated with feedback.

On Thu, Jun 1, 2017 at 6:56 PM, Arthur Evans <notifications@github.com>
wrote:

> *@arthurevans* commented on this pull request.
>
> Left feedback.
> Also, need to add nav entries for the new stuff.
> ------------------------------
>
> In app/2.0/docs/tools/create-app-polymer-cli.md
> <https://github.com/Polymer/docs/pull/2130#discussion_r119759669>:
>
> > @@ -0,0 +1,101 @@
> +---
> +title: Create an application project with the Polymer CLI
> +---
> +
> +<!-- toc -->
> +
> +Polymer CLI supports initializing a project folder with one of several application templates.
> +The `polymer-2-application` template is the most basic starting point for a Polymer-based
> +application. Other templates introduce more complex layout and application patterns.
>
> I would add a specific callout here for the PSK template. Like,
>
> "For a more full-featured progressive web app template, you can use the
> starter kit template (polymer-2-starter-kit). See [link] for a tutorial
> on setting up the starter kit.
> ------------------------------
>
> In app/2.0/docs/tools/create-element-polymer-cli.md
> <https://github.com/Polymer/docs/pull/2130#discussion_r119761090>:
>
> > @@ -0,0 +1,121 @@
> +---
> +title: Create an element project with the Polymer CLI
> +---
> +
> +<!-- toc -->
> +
> +This section shows you how to start an element project.
> +
> +1.  Create a directory for your element project. It's best practice for the name
>
> @FredKSchott <https://github.com/fredkschott> Could we say "common"?
> Like, "For projects with a single element, it's common to name the project
> directory to match the element name."
> ------------------------------
>
> In app/2.0/docs/tools/create-app-polymer-cli.md
> <https://github.com/Polymer/docs/pull/2130#discussion_r119763815>:
>
> > +    Source code for main element.
> +*    `test/`<code><var>my-app</var></code>/<code><var>my-app</var></code>`_test.html`. Tests for
> +main element.
> +
> +### Add elements
> +
> +You may want to compose your main element out of smaller, application-specific elements. These
> +app-specific elements should be defined in the `src` directory, at the same level as
> +<code><var>my-app</var></code>.
> +
> +    app/
> +      src/
> +        my-app/
> +        my-el/
> +
> +Currently there is no Polymer CLI command to generate application-specific elements. You should do
>
> I think we kind of need to highlight this to prevent people from trying to
> do init element inside an app project. But instead of saying, "There is
> currently no..." We could say,
>
> To add another element to the project:
>
>    1.
>
>    Create a new folder under src.
>
>     mkdir src/my-el
>
>    2.
>
>    Create an HTML import for the new element. You can use the existing
>    app element as a starting point.
>    3.
>
>    To use the new element, you'll need to import it into your application
>    element.
>
> Don't use polymer init to create an element project inside your app
> project.
>
> This adds more detail without being entirely satisfying (i.e., doesn't
> show adding the <link rel="import"> or using the element in the app
> element). So alternately, just add the .html files to the diagram above for
> clarity:
>
> my-example/
>      src/
>        my-example-app/
>          my-example-app.html
>        my-el/
>          my-el.html
>
> You can copy the existing app element as a starting point for your next
> element.
>
> Don't use polymer init to create an element project inside your app
> project.
>
> === end suggestion
>
> I changed the example here to "my-example" so the directory names would
> make sense. If you're using my-app as a placeholder, should be
> <var>my-app</var> by style. Downside: examples might need to be formatted
> in HTML instead of fenced code blocks, because fenced code blocks escape
> the HTML markup.
> ------------------------------
>
> In app/2.0/docs/tools/create-element-polymer-cli.md
> <https://github.com/Polymer/docs/pull/2130#discussion_r119764156>:
>
> > +
> +*   `bower.json`. Configuration file for Bower.
> +*   `bower_components/`. Project dependencies. See [Manage dependencies](#dependencies).
> +*   `demo/index.html`. Demo of <code><var>my-el</var></code>`.html`.
> +*   `index.html`. Automatically-generated API reference.
> +*   <code><var>my-el</var></code>`.html`. Element source code.
> +*   `test/`<code><var>my-el</var></code>`_test.html`. Unit tests for
> +    the element.
> +
> +## Manage dependencies and HTML imports {#dependencies}
> +
> +### Use Bower to manage dependencies
> +
> +Polymer CLI uses [Bower](http://bower.io) for dependency management.
> +
> +Dependencies are stored in the `bower_components` directory. You should never manually alter the
>
> Since we're just about the only folks who use it, seems like it's probably
> OK. If you're comfortable with Bower and CLI tools, chances are you aren't
> even reading the "create an element project/app project" docs. I wouldn't
> want to give a full tutorial, but a few "don't stick your fork in this here
> socket" pointers seem apropos?
> ------------------------------
>
> In app/2.0/docs/tools/create-element-polymer-cli.md
> <https://github.com/Polymer/docs/pull/2130#discussion_r119765585>:
>
> > +*   `demo/index.html`. Demo of <code><var>my-el</var></code>`.html`.
> +*   `index.html`. Automatically-generated API reference.
> +*   <code><var>my-el</var></code>`.html`. Element source code.
> +*   `test/`<code><var>my-el</var></code>`_test.html`. Unit tests for
> +    the element.
> +
> +## Manage dependencies and HTML imports {#dependencies}
> +
> +### Use Bower to manage dependencies
> +
> +Polymer CLI uses [Bower](http://bower.io) for dependency management.
> +
> +Dependencies are stored in the `bower_components` directory. You should never manually alter the
> +contents of this directory.
> +
> +Use the Bower CLI to manage dependencies.
>
> This is a good question. Don't want to hang up on it too much here (we use
> bower directly other places), but we should discuss whether to move to that
> in general (or offer both options).
> ------------------------------
>
> In app/2.0/docs/tools/create-element-polymer-cli.md
> <https://github.com/Polymer/docs/pull/2130#discussion_r119765766>:
>
> > +*   <code><var>my-el</var></code>`.html`. Element source code.
> +*   `test/`<code><var>my-el</var></code>`_test.html`. Unit tests for
> +    the element.
> +
> +## Manage dependencies and HTML imports {#dependencies}
> +
> +### Use Bower to manage dependencies
> +
> +Polymer CLI uses [Bower](http://bower.io) for dependency management.
> +
> +Dependencies are stored in the `bower_components` directory. You should never manually alter the
> +contents of this directory.
> +
> +Use the Bower CLI to manage dependencies.
> +
> +This command downloads a dependency to `bower_components/` (the `--save` flag saves the new
>
> Rather than "This command" form, maybe replace these with:
>
> To add a new Bower dependency to your project:
>
> <bower command>
>
> As Fred suggested, "This command..." is a little ambiguous about whether
> you're talking about the command above, or the command below.
> ------------------------------
>
> In app/2.0/docs/tools/polymer-cli-commands.md
> <https://github.com/Polymer/docs/pull/2130#discussion_r119766666>:
>
> > +* [`polymer build` (for app projects only)](#build)
> +* [`polymer init`](#init)
> +* [`polymer install`](#install)
> +* [`polymer lint`](#lint)
> +* [`polymer serve`](#serve)
> +* [`polymer test`](#tests)
> +* [Global options for Polymer CLI commands](#global)
> +
> +## `polymer build` {#build}
> +
> +*This command is for app projects only.*
> +
> +Generates a production-ready build of your app. This process includes minifying the HTML, CSS, and
> +JS of the application dependencies, and generating a service worker to pre-cache dependencies.
> +
> +Polymer CLI's build process is designed for apps that follow the [app shell
>
> I think this link is a little confusing because some of our app structure
> is Polymer- or PRPL-specific (Like, the terms "entrypoint" and "fragment").
> And that doc is full of a lot of implementation details that won't be
> helpful to the average Polymer user.
>
> It would make more sense to link here to the PRPL doc, and have only that
> doc link to the app shell article for background on the app shell pattern.
> ------------------------------
>
> In app/2.0/docs/tools/polymer-cli.md
> <https://github.com/Polymer/docs/pull/2130#discussion_r119767221>:
>
> > @@ -39,500 +39,28 @@ You're all set. Run `polymer help` to view a list of commands.
>
>  ## Overview {#overview}
>
> -Polymer CLI is a command-line interface for Polymer projects. It includes a build pipeline, a
> -boilerplate generator for creating elements and apps, a linter, a development server, and a test
> -runner.
> +Polymer CLI is our official command line tool for Polymer projects and Web Components. It includes
>
> our => the (see style/person)
> ------------------------------
>
> In app/2.0/docs/tools/polymer-cli.md
> <https://github.com/Polymer/docs/pull/2130#discussion_r119767461>:
>
> > @@ -39,500 +39,28 @@ You're all set. Run `polymer help` to view a list of commands.
>
>  ## Overview {#overview}
>
> I know this organization is not new, but it strikes me as strange that the
> overview follows the install.
>
> Maybe move the first sentence to the head of the article (before the
> Install heading) and retitle this section "Types of projects" or "CLI
> project types".
> ------------------------------
>
> In app/2.0/docs/tools/polymer-cli.md
> <https://github.com/Polymer/docs/pull/2130#discussion_r119767531>:
>
> > @@ -39,500 +39,28 @@ You're all set. Run `polymer help` to view a list of commands.
>
>  ## Overview {#overview}
>
> And maybe ## Install => ## Install Polymer CLI?
>
> Not sure. Do what feels right.
> ------------------------------
>
> In app/2.0/docs/tools/polymer-cli-commands.md
> <https://github.com/Polymer/docs/pull/2130#discussion_r119767582>:
>
> > +
> +This section explains various useful Polymer CLI commands that you'll want to incorporate into your
> +development workflow while you build your element or app project.
> +
> +The commands are intended for both element and app projects unless otherwise
> +noted.
> +
> +* [`polymer build` (for app projects only)](#build)
> +* [`polymer init`](#init)
> +* [`polymer install`](#install)
> +* [`polymer lint`](#lint)
> +* [`polymer serve`](#serve)
> +* [`polymer test`](#tests)
> +* [Global options for Polymer CLI commands](#global)
> +
> +## `polymer build` {#build}
>
> As discussed, remove code formatting in headers.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Polymer/docs/pull/2130#pullrequestreview-41669870>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AVLM4cQ_Klj6f7StqhVTD1nU2mLEhqGEks5r_2uxgaJpZM4Nb3WB>
> .
>

katejeffreys commented:
Updated nav. Staged here:
https://restructure-cli-docs-dot-polymer-project.appspot.com/2.0/docs/devguide/feature-overview


On Fri, Jun 2, 2017 at 2:54 PM, Kate Jeffreys <katejeffreys@google.com>
wrote:

> Thanks, updated with feedback.
>
> On Thu, Jun 1, 2017 at 6:56 PM, Arthur Evans <notifications@github.com>
> wrote:
>
>> *@arthurevans* commented on this pull request.
>>
>> Left feedback.
>> Also, need to add nav entries for the new stuff.
>> ------------------------------
>>
>> In app/2.0/docs/tools/create-app-polymer-cli.md
>> <https://github.com/Polymer/docs/pull/2130#discussion_r119759669>:
>>
>> > @@ -0,0 +1,101 @@
>> +---
>> +title: Create an application project with the Polymer CLI
>> +---
>> +
>> +<!-- toc -->
>> +
>> +Polymer CLI supports initializing a project folder with one of several application templates.
>> +The `polymer-2-application` template is the most basic starting point for a Polymer-based
>> +application. Other templates introduce more complex layout and application patterns.
>>
>> I would add a specific callout here for the PSK template. Like,
>>
>> "For a more full-featured progressive web app template, you can use the
>> starter kit template (polymer-2-starter-kit). See [link] for a tutorial
>> on setting up the starter kit.
>> ------------------------------
>>
>> In app/2.0/docs/tools/create-element-polymer-cli.md
>> <https://github.com/Polymer/docs/pull/2130#discussion_r119761090>:
>>
>> > @@ -0,0 +1,121 @@
>> +---
>> +title: Create an element project with the Polymer CLI
>> +---
>> +
>> +<!-- toc -->
>> +
>> +This section shows you how to start an element project.
>> +
>> +1.  Create a directory for your element project. It's best practice for the name
>>
>> @FredKSchott <https://github.com/fredkschott> Could we say "common"?
>> Like, "For projects with a single element, it's common to name the project
>> directory to match the element name."
>> ------------------------------
>>
>> In app/2.0/docs/tools/create-app-polymer-cli.md
>> <https://github.com/Polymer/docs/pull/2130#discussion_r119763815>:
>>
>> > +    Source code for main element.
>> +*    `test/`<code><var>my-app</var></code>/<code><var>my-app</var></code>`_test.html`. Tests for
>> +main element.
>> +
>> +### Add elements
>> +
>> +You may want to compose your main element out of smaller, application-specific elements. These
>> +app-specific elements should be defined in the `src` directory, at the same level as
>> +<code><var>my-app</var></code>.
>> +
>> +    app/
>> +      src/
>> +        my-app/
>> +        my-el/
>> +
>> +Currently there is no Polymer CLI command to generate application-specific elements. You should do
>>
>> I think we kind of need to highlight this to prevent people from trying
>> to do init element inside an app project. But instead of saying, "There is
>> currently no..." We could say,
>>
>> To add another element to the project:
>>
>>    1.
>>
>>    Create a new folder under src.
>>
>>     mkdir src/my-el
>>
>>    2.
>>
>>    Create an HTML import for the new element. You can use the existing
>>    app element as a starting point.
>>    3.
>>
>>    To use the new element, you'll need to import it into your
>>    application element.
>>
>> Don't use polymer init to create an element project inside your app
>> project.
>>
>> This adds more detail without being entirely satisfying (i.e., doesn't
>> show adding the <link rel="import"> or using the element in the app
>> element). So alternately, just add the .html files to the diagram above for
>> clarity:
>>
>> my-example/
>>      src/
>>        my-example-app/
>>          my-example-app.html
>>        my-el/
>>          my-el.html
>>
>> You can copy the existing app element as a starting point for your next
>> element.
>>
>> Don't use polymer init to create an element project inside your app
>> project.
>>
>> === end suggestion
>>
>> I changed the example here to "my-example" so the directory names would
>> make sense. If you're using my-app as a placeholder, should be
>> <var>my-app</var> by style. Downside: examples might need to be
>> formatted in HTML instead of fenced code blocks, because fenced code blocks
>> escape the HTML markup.
>> ------------------------------
>>
>> In app/2.0/docs/tools/create-element-polymer-cli.md
>> <https://github.com/Polymer/docs/pull/2130#discussion_r119764156>:
>>
>> > +
>> +*   `bower.json`. Configuration file for Bower.
>> +*   `bower_components/`. Project dependencies. See [Manage dependencies](#dependencies).
>> +*   `demo/index.html`. Demo of <code><var>my-el</var></code>`.html`.
>> +*   `index.html`. Automatically-generated API reference.
>> +*   <code><var>my-el</var></code>`.html`. Element source code.
>> +*   `test/`<code><var>my-el</var></code>`_test.html`. Unit tests for
>> +    the element.
>> +
>> +## Manage dependencies and HTML imports {#dependencies}
>> +
>> +### Use Bower to manage dependencies
>> +
>> +Polymer CLI uses [Bower](http://bower.io) for dependency management.
>> +
>> +Dependencies are stored in the `bower_components` directory. You should never manually alter the
>>
>> Since we're just about the only folks who use it, seems like it's
>> probably OK. If you're comfortable with Bower and CLI tools, chances are
>> you aren't even reading the "create an element project/app project" docs. I
>> wouldn't want to give a full tutorial, but a few "don't stick your fork in
>> this here socket" pointers seem apropos?
>> ------------------------------
>>
>> In app/2.0/docs/tools/create-element-polymer-cli.md
>> <https://github.com/Polymer/docs/pull/2130#discussion_r119765585>:
>>
>> > +*   `demo/index.html`. Demo of <code><var>my-el</var></code>`.html`.
>> +*   `index.html`. Automatically-generated API reference.
>> +*   <code><var>my-el</var></code>`.html`. Element source code.
>> +*   `test/`<code><var>my-el</var></code>`_test.html`. Unit tests for
>> +    the element.
>> +
>> +## Manage dependencies and HTML imports {#dependencies}
>> +
>> +### Use Bower to manage dependencies
>> +
>> +Polymer CLI uses [Bower](http://bower.io) for dependency management.
>> +
>> +Dependencies are stored in the `bower_components` directory. You should never manually alter the
>> +contents of this directory.
>> +
>> +Use the Bower CLI to manage dependencies.
>>
>> This is a good question. Don't want to hang up on it too much here (we
>> use bower directly other places), but we should discuss whether to move to
>> that in general (or offer both options).
>> ------------------------------
>>
>> In app/2.0/docs/tools/create-element-polymer-cli.md
>> <https://github.com/Polymer/docs/pull/2130#discussion_r119765766>:
>>
>> > +*   <code><var>my-el</var></code>`.html`. Element source code.
>> +*   `test/`<code><var>my-el</var></code>`_test.html`. Unit tests for
>> +    the element.
>> +
>> +## Manage dependencies and HTML imports {#dependencies}
>> +
>> +### Use Bower to manage dependencies
>> +
>> +Polymer CLI uses [Bower](http://bower.io) for dependency management.
>> +
>> +Dependencies are stored in the `bower_components` directory. You should never manually alter the
>> +contents of this directory.
>> +
>> +Use the Bower CLI to manage dependencies.
>> +
>> +This command downloads a dependency to `bower_components/` (the `--save` flag saves the new
>>
>> Rather than "This command" form, maybe replace these with:
>>
>> To add a new Bower dependency to your project:
>>
>> <bower command>
>>
>> As Fred suggested, "This command..." is a little ambiguous about whether
>> you're talking about the command above, or the command below.
>> ------------------------------
>>
>> In app/2.0/docs/tools/polymer-cli-commands.md
>> <https://github.com/Polymer/docs/pull/2130#discussion_r119766666>:
>>
>> > +* [`polymer build` (for app projects only)](#build)
>> +* [`polymer init`](#init)
>> +* [`polymer install`](#install)
>> +* [`polymer lint`](#lint)
>> +* [`polymer serve`](#serve)
>> +* [`polymer test`](#tests)
>> +* [Global options for Polymer CLI commands](#global)
>> +
>> +## `polymer build` {#build}
>> +
>> +*This command is for app projects only.*
>> +
>> +Generates a production-ready build of your app. This process includes minifying the HTML, CSS, and
>> +JS of the application dependencies, and generating a service worker to pre-cache dependencies.
>> +
>> +Polymer CLI's build process is designed for apps that follow the [app shell
>>
>> I think this link is a little confusing because some of our app structure
>> is Polymer- or PRPL-specific (Like, the terms "entrypoint" and "fragment").
>> And that doc is full of a lot of implementation details that won't be
>> helpful to the average Polymer user.
>>
>> It would make more sense to link here to the PRPL doc, and have only that
>> doc link to the app shell article for background on the app shell pattern.
>> ------------------------------
>>
>> In app/2.0/docs/tools/polymer-cli.md
>> <https://github.com/Polymer/docs/pull/2130#discussion_r119767221>:
>>
>> > @@ -39,500 +39,28 @@ You're all set. Run `polymer help` to view a list of commands.
>>
>>  ## Overview {#overview}
>>
>> -Polymer CLI is a command-line interface for Polymer projects. It includes a build pipeline, a
>> -boilerplate generator for creating elements and apps, a linter, a development server, and a test
>> -runner.
>> +Polymer CLI is our official command line tool for Polymer projects and Web Components. It includes
>>
>> our => the (see style/person)
>> ------------------------------
>>
>> In app/2.0/docs/tools/polymer-cli.md
>> <https://github.com/Polymer/docs/pull/2130#discussion_r119767461>:
>>
>> > @@ -39,500 +39,28 @@ You're all set. Run `polymer help` to view a list of commands.
>>
>>  ## Overview {#overview}
>>
>> I know this organization is not new, but it strikes me as strange that
>> the overview follows the install.
>>
>> Maybe move the first sentence to the head of the article (before the
>> Install heading) and retitle this section "Types of projects" or "CLI
>> project types".
>> ------------------------------
>>
>> In app/2.0/docs/tools/polymer-cli.md
>> <https://github.com/Polymer/docs/pull/2130#discussion_r119767531>:
>>
>> > @@ -39,500 +39,28 @@ You're all set. Run `polymer help` to view a list of commands.
>>
>>  ## Overview {#overview}
>>
>> And maybe ## Install => ## Install Polymer CLI?
>>
>> Not sure. Do what feels right.
>> ------------------------------
>>
>> In app/2.0/docs/tools/polymer-cli-commands.md
>> <https://github.com/Polymer/docs/pull/2130#discussion_r119767582>:
>>
>> > +
>> +This section explains various useful Polymer CLI commands that you'll want to incorporate into your
>> +development workflow while you build your element or app project.
>> +
>> +The commands are intended for both element and app projects unless otherwise
>> +noted.
>> +
>> +* [`polymer build` (for app projects only)](#build)
>> +* [`polymer init`](#init)
>> +* [`polymer install`](#install)
>> +* [`polymer lint`](#lint)
>> +* [`polymer serve`](#serve)
>> +* [`polymer test`](#tests)
>> +* [Global options for Polymer CLI commands](#global)
>> +
>> +## `polymer build` {#build}
>>
>> As discussed, remove code formatting in headers.
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/Polymer/docs/pull/2130#pullrequestreview-41669870>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AVLM4cQ_Klj6f7StqhVTD1nU2mLEhqGEks5r_2uxgaJpZM4Nb3WB>
>> .
>>
>
>

arthurevans commented on app/2.0/docs/tools/create-app-polymer-cli.md:
```diff
@@ -0,0 +1,124 @@
+See [Polymer App Toolbox templates](../../toolbox/templates) for more details on other templates.
+
+For a more full-featured progressive web app template, you can use the starter kit template 
+(`polymer-2-starter-kit`). See [/{{{polymer_version_dir}}}/start/toolbox/set-up](the Polymer Starter Kit tutorial)
```
lInk is backwards, should be `[text](url)`

arthurevans commented on app/2.0/docs/tools/create-app-polymer-cli.md:
```diff
@@ -0,0 +1,124 @@
+application-specific elements should be defined in the `src` directory, at the same level as 
+<code><var>my-app</var></code>.
+
+<code>
```
Needs to be wrapped in a `<pre>` tag, otherwise the formatting gets lost.

arthurevans commented on app/2.0/docs/tools/create-app-polymer-cli.md:
```diff
@@ -0,0 +1,124 @@
+
+1. Create a new folder under src.
+   
+<code>
```
This is getting outdented relative to the list. I _think_ it'll work OK if you intend this whole block four spaces.

arthurevans commented on app/2.0/docs/tools/create-app-polymer-cli.md:
```diff
@@ -0,0 +1,124 @@
+3. To use the new element, you'll need to import it into your application element (for example, <code><var>my-app</var>.html</code>) with an "import" link:
+   
+<code>
+   <link rel="import" href="/src/<var>my-el</var>/<var>my-el</var>.html">
```
Again, indent four spaces. Plus, you'll need to replace the `<` and `>` with entities: `&gt;` in place of `<` and `&lt;` in place of `>` ... The markdown blocks escape these automatically. When you do a `<code>` block yourself, you need to escape them or this gets interpreted as a literal `<link>` tag and it is hidden from the user.

arthurevans commented on app/2.0/docs/tools/create-app-polymer-cli.md:
```diff
@@ -0,0 +1,124 @@
+    <!-- from a 'src/some-application.html' file in your application -->
+    <link rel="import" href="../bower_components/polymer/polymer.html">
+
+Only one file should use absolute URLs for imports: your "entrypoint" (default: "index.html"). 
```
I think we can delete these two paragraphs (this and the following). Although the current basic app template uses absolute URLs, PSK now uses relative URLs in the entrypoint and I think we'll be going that way in the future. So we probably don't need to call this out.

katejeffreys replied:
Could you point out the paragraphs that need to be removed by their starting words, please? I can't disambiguate.

arthurevans replied:
"Only one file should use absolute URLs for imports: your "entrypoint" (default: "index.html").  ... and the following code sample."

arthurevans replied:
In other words, from L100 on.

arthurevans commented on app/2.0/nav.yaml:
```diff
@@ -133,6 +133,15 @@
   - title: Polymer CLI
     path: /2.0/docs/tools/polymer-cli
     indent: True
+  - title: Polymer CLI commands
```
The left nav looks pretty cluttered with all of the "Polymer CLIs" in there... Maybe we could shorten these three to:
CLI commands
Create an element project
Create an app project

(Just in the left nav. Full titles in the docs are a-ok.)

arthurevans commented on app/2.0/docs/tools/polymer-cli.md:
```diff
@@ -4,14 +4,18 @@ title: Polymer CLI
+
+## Install Polymer CLI {#install}
 
 1.  Make sure you have installed a verson of Node.js supported by Polymer. To check the version
```
<typo>

arthurevans commented on app/2.0/docs/tools/polymer-cli-commands.md:
```diff
@@ -0,0 +1,277 @@
+
+    polymer lint --rules=polymer-2
+
+This will lint all of the code in your project with the `polymer-2` ruleset, which is appropriate for projects using Polymer 2.0. If your code is hybrid and should with with either Polymer 1.x or 2.x then `polymer-2-hybrid` is a better choice, as it will warn you about use of features that do not exist in Polymer 2.x.
```
repeated word (with with)
